contracts/beyondFramework/dca/base/DCABaseUpgradeable.sol

```diff
@@ -0,0 +1,1168 @@
+                // since protocol might mint less or more tokens refresh amount
+                receivedBluechip = _invest(receivedBluechip);
+            }
+
+            // store information about spent asset and received asset
+            dcaHistory.addHistoricalGauge(depositedAmount, receivedBluechip);
+
+            // compute totals for event
+            totalDepositSpent += depositedAmount;
+            totalBluechipReceived += receivedBluechip;
+        }
+
+        if (bluechipInvestmentState == BluechipInvestmentState.Investing) {
+            // claim rewards
+            uint256 claimedBluechipRewards = _claimRewards();
+
+            // if something was claimed invest rewards and increase current gauge
+            if (claimedBluechipRewards > 0) {
+                claimedBluechipRewards = _invest(claimedBluechipRewards);
+                dcaHistory.increaseGaugeAt(
+                    claimedBluechipRewards,
+                    dcaHistory.currentHistoricalIndex() - 1
+                );
+
+                // increase total amount for event
+                totalBluechipReceived += claimedBluechipRewards;
+            }
+        }
+
+        // update last invest timestamp
+        lastInvestmentTimestamp += passedInvestPeriods * investmentPeriod;
+
+        emit Invest(
+            totalDepositSpent,
+            totalBluechipReceived,
+            lastInvestmentTimestamp
+        );
+    }
+
+    function withdrawAll(bool convertBluechipIntoDepositAsset)
+        public
+        virtual
+        nonReentrant
+        whenNotPaused
+        emergencyWithdrawOnEmergencyExitedStatus(_msgSender())
+    {
+        _withdrawAll(_msgSender(), convertBluechipIntoDepositAsset);
+    }
+
+    function withdrawAllFor(
+        address sender,
+        bool convertBluechipIntoDepositAsset
+    )
+        public
+        virtual
+        onlyPortfolio
+        nonReentrant
+        whenNotPaused
+        emergencyWithdrawOnEmergencyExitedStatus(sender)
+    {
+        _withdrawAll(sender, convertBluechipIntoDepositAsset);
+    }
+
+    function _withdrawAll(address sender, bool convertBluechipIntoDepositAsset)
+        private
+    {
+        // define total not invested yet amount by user
+        // and total bought bluechip asset amount
+        uint256 notInvestedYet;
+        uint256 investedIntoBluechip;
+
+        DcaDepositor storage depositor = depositors[sender];
+        for (uint256 i = 0; i < depositor.positions.length; i++) {
+            (
+                uint256 positionBluechipInvestment,
+                uint256 positionNotInvestedYet
+            ) = _computePositionWithdrawAll(depositor.positions[i]);
+
+            // increase users total amount
+            investedIntoBluechip += positionBluechipInvestment;
+            notInvestedYet += positionNotInvestedYet;
+        }
+
+        // since depositor withdraws everything
+        // we can remove his data completely
+        delete depositors[sender];
```

Comment:
this delete's gas cost is linear in terms of the existing positions for that user

do we already know up to how many positions this is safe?